archive/imo/imo1979_q1.lean

```diff
@@ -0,0 +1,250 @@
+/-
+Copyright (c) 2020 Kevin Buzzard. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kevin Buzzard
+-/
+import algebra.big_operators.ring
+import data.nat.interval
+import data.nat.parity
+import number_theory.padics.padic_norm
+import tactic.linarith
+import tactic.field_simp
+
+/-!
+# IMO 1979 Q1
+
+Let `p` and `q` be positive integers such that
+
+`p/q = 1 - 1/2 + 1/3 - 1/4 + ... - 1/1318 + 1/1319`
+
+Prove that `p` is a multiple of `1979`.
+
+## The solution
+
+The proof we formalise is the following. Rewrite the sum as
+`1 + 1/2 + 1/3 + ... + 1/1319 - 2 * (1/2 + 1/4 + 1/6 + ... + 1/1318) = 1/660 + 1/661 + ... + 1/1319`
+and now re-arranging as `(1/660 + 1/1319) + (1/661 + 1/1318) + ...` we see that
+the numerator of each fraction is `1979` and the denominator is coprime to `1979`
+(note that `1979` is prime). Hence the `1979`-adic valuation of each fraction is positive,
+and thus the `1979`-adic valuation of the sum is also positive.
+
+## Remarks on the formalisation
+
+The p-adic valuation function on ℚ is ℤ-valued and hence has v(0)=0 (rather than +∞);
+we thus had to occasionally deal with funny edge cases which aren't there mathematically.
+In retrospect it might have been easier to work with p-adic norms, which are ℚ-valued
```

Comment:
How hard would it be to do this refactor? I think these IMO files should be showcases that look as nice as possible.

Reply:
@kbuzzard, what did you exactly mean by that? I can't make sense of the "`ℚ`-valued" bit. Surely you meant "`with_top ℤ`-valued"? If that's really the case, the refactor is arguable.

Reply:
I think he meant `ℝ`-valued, although the norms of p-adic numbers are all in the image of `coe : ℚ → ℝ`.

Reply:
I meant that valuations have problems with v(0) but norms don't, |0|=0. We have |p^n|=p^{-n} so \Q will do.